pypicloud/access/_ldap.py

```diff
@@ -0,0 +1,391 @@
+"""LDAP authentication plugin for pypicloud."""
+
+
+try:
+    import ldap  # pylint: disable=F0401
+except ImportError:
+    LDAP_ENABLED = False
+else:
+    LDAP_ENABLED = True
+
+import logging
+from functools import wraps
+
+from .base import IAccessBackend
+
+
+def reconnect(func):
+    """If the LDAP connection dies underneath us, recreate it."""
+
+    @wraps(func)
+    def _reconnect(*args, **kwargs):
+        """Inner wrap function to reconnect on failure."""
+
+        try:
+            return func(*args, **kwargs)
+        except ldap.LDAPError:
+            LDAP._connect()
+            return func(*args, **kwargs)
+
+    return _reconnect
+
+
+class LDAP(object):
+    """Handles interactions with the remote LDAP server."""
+
+    @staticmethod
+    def configure(settings=None):
+        """Configures self with the settings dictionary."""
+
+        LDAP._id_field = settings["auth.ldap.id_field"]
+        LDAP._url = settings["auth.ldap.url"]
+        LDAP._service_dn = settings["auth.ldap.service_dn"]
+        LDAP._service_password = settings["auth.ldap.service_password"]
+        LDAP._base_dn = settings["auth.ldap.base_dn"]
+        LDAP._all_user_search = settings["auth.ldap.all_user_search"]
+        LDAP._admin_field = settings["auth.ldap.admin_field"]
+        LDAP._admin_dns = [
+            dn for dn in settings["auth.ldap.admin_dns"].splitlines() if dn
+        ]
+
+        LDAP._connect()
+        LDAP.all_users()
+
+    @staticmethod
+    def _connect():
+        """Initializes the python-ldap module and does the initial bind."""
+
+        LDAP._server = ldap.initialize(LDAP._url)
+        LDAP._server.simple_bind_s(LDAP._service_dn, LDAP._service_password)
+
+    @staticmethod
+    @reconnect
```

Comment:
basically any method that deals with the LDAP._server object, besides the initial one as we can be pretty sure it's not going to die. ldap connections are usually fairly short lived though, so just to auto rebind whenever that happens